lib/src/value/number.ts

```diff
@@ -0,0 +1,718 @@
+  frequency: ['Hz', 'kHz'],
+  'pixel density': ['dpi', 'dpcm', 'dppx'],
+};
+
+// A map from each unit to its human-readable type.
+const typesByUnit: Record<string, string> = {};
+for (const [type, units] of Object.entries(unitsByType)) {
+  for (const unit of units) {
+    typesByUnit[unit] = type;
+  }
+}
+
+/** A SassScript number. */
+export class SassNumber extends Value {
+  private numeratorUnitsInternal: List<string>;
+  private denominatorUnitsInternal: List<string>;
+
+  /** Creates a number, optionally with a single numerator unit. */
+  constructor(private readonly valueInternal: number, unit?: string) {
+    super();
+    this.numeratorUnitsInternal = unit === undefined ? List([]) : List([unit]);
+    this.denominatorUnitsInternal = List([]);
+    Object.freeze(this);
+  }
+
+  /**
+   * Creates a number with `numeratorUnits` and `denominatorUnits`.
+   *
+   * Upon construction, any compatible numerator and denominator units are
+   * simplified away according to the conversion factor between them.
+   */
+  static withUnits(
+    value: number,
+    options?: {
+      numeratorUnits?: string[] | List<string>;
+      denominatorUnits?: string[] | List<string>;
+    }
+  ): SassNumber {
+    let numerators = asImmutableList(options?.numeratorUnits ?? []);
+    const unsimplifiedDenominators = options?.denominatorUnits ?? [];
+
+    const denominators = [];
+    for (const denominator of unsimplifiedDenominators) {
+      let simplifiedAway = false;
+      for (const [i, numerator] of numerators.entries()) {
+        const factor = conversionFactor(denominator, numerator);
+        if (factor === null) continue;
+        value /= factor;
+        numerators = numerators.delete(i);
+        simplifiedAway = true;
+        break;
+      }
+      if (!simplifiedAway) denominators.push(denominator);
+    }
+
+    const number = new SassNumber(value);
+    number.numeratorUnitsInternal = numerators;
+    number.denominatorUnitsInternal = List(denominators);
+    return number;
+  }
+
+  /** `this`'s value. */
+  get value(): number {
+    return this.valueInternal;
+  }
+
+  /** Whether `value` is an integer. */
+  get isInt(): boolean {
+    return fuzzyIsInt(this.value);
+  }
+
+  /**
+   * If `value` is an integer according to `isInt`, returns it as an integer.
+   *
+   * Otherwise, returns null.
+   */
+  get asInt(): number | null {
+    return fuzzyAsInt(this.value);
+  }
+
+  /** `this`'s numerator units. */
+  get numeratorUnits(): List<string> {
+    return this.numeratorUnitsInternal;
+  }
+
+  /** `this`'s denominator units. */
+  get denominatorUnits(): List<string> {
+    return this.denominatorUnitsInternal;
+  }
+
+  /** Whether `this` has any units. */
+  get hasUnits(): boolean {
+    return !(this.numeratorUnits.isEmpty() && this.denominatorUnits.isEmpty());
+  }
+
+  assertNumber(): SassNumber {
+    return this;
+  }
+
+  /**
+   * If `value` is an integer according to `isInt`, returns it as an integer.
```

Comment:
"as an integer" is a bit ambiguous since JS doesn't have a proper integer type. Maybe "rounded to that integer"?